SimTKcommon/src/ParallelExecutor.cpp

```diff
@@ -25,23 +25,36 @@
 #include "SimTKcommon/internal/ParallelExecutor.h"
 #include <pthread.h>
 
+#include <iostream>
+#include <string>
+#include <algorithm>
+
+using namespace std;
+
 namespace SimTK {
 
 static void* threadBody(void* args);
 
+ParallelExecutorImpl::ParallelExecutorImpl() : finished(false) {
+
+    // Set the maximum number of threads that we can use
+    SimTK_APIARGCHECK_ALWAYS(ParallelExecutor::getNumProcessors() > 0, "ParallelExecutorImpl", "ParallelExecutorImpl", "Number of threads detected is not positive.");
+    NUM_MAX_THREADS = ParallelExecutor::getNumProcessors();
+
+    pthread_mutex_init(&runLock, NULL);
+    pthread_cond_init(&runCondition, NULL);
+    pthread_cond_init(&waitCondition, NULL);
+
+}
 ParallelExecutorImpl::ParallelExecutorImpl(int numThreads) : finished(false) {
 
-    // Construct all the threading related objects we will need.
-    
+    // Set the maximum number of threads that we can use
     SimTK_APIARGCHECK_ALWAYS(numThreads > 0, "ParallelExecutorImpl", "ParallelExecutorImpl", "Number of threads must be positive.");
-    threads.resize(numThreads);
+    NUM_MAX_THREADS = numThreads;
+
     pthread_mutex_init(&runLock, NULL);
     pthread_cond_init(&runCondition, NULL);
     pthread_cond_init(&waitCondition, NULL);
-    for (int i = 0; i < numThreads; ++i) {
-        threadInfo.push_back(new ThreadInfo(i, this));
-        pthread_create(&threads[i], NULL, threadBody, threadInfo[i]);
-    }
 }
```

Comment:
I delayed thread construction from the constructor to the execute() method because if all of the forces were non-parallel but the ParallelExecutor was constructed with default (empty) parameters it would spend time launching approx. 8 threads. I don't see any added safety launching the threads at construction vs execution. Thoughts?